### PR TITLE
Rebinding keys on mices and keyboards

### PR DESCRIPTION
**Pull request by pmizio**
Hello I created some code for rebinding, I hope it's ok and I'm asking for strict code review because it's my first low level code since my gaduation(3 of 4 years ago).

Added new dependency to convert keycodes int <-> string - python-evdev

Tested on my Arch box(5.2.1-arch1-1-ARCH)
Hardware:
 - Razer Naga Chroma
 - Razer Orbweaver Chroma

Cheers

### CODE REVIEW

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -2407,6 +2443,9 @@ static void razer_kbd_disconnect(struct hid_device *hdev)
             break;
         }
     } else if(intf->cur_altsetting->desc.bInterfaceProtocol == USB_INTERFACE_PROTOCOL_KEYBOARD) {
+        // Translations
+        device_remove_file(&hdev->dev, &dev_attr_button_translations);
+
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:2592]: Is there a `razer_cleanup_translations` call missing in this file?

**Comment by z3ntu** [comment]
Really good code, I'm impressed :)

I'll do some proper code review + testing in the coming days but it looks good from a quick overview - but I definitely know less about input in the kernel than you from looking at the PR.

File: driver/razerkbd_driver.c (hunk 1 above)
Comment (pmizio) [reply driver/razerkbd_driver.c:2592]: You're right i forgot to move it after develop in mouse driver ;)

**Comment by tildearrow** [comment]
Minor nitpick: This may cause issues with the Ripple effect.

**Comment by pmizio** [comment]
> Minor nitpick: This may cause issues with the Ripple effect.

My devices don't support ripple effect so sadly I cannot test this case.

**Comment by devclef** [comment]
Hi, 

I got this working on my PC as well.. (Probably not the cleanest way, but using your forked repo and modifying the PKGBUILD from AUR to use it as a source)

Orbweaver Chroma
Arch  - Linux 5.2.7.arch1-1

One question/suggestion:
Either define the keymaps (KEY_1, etc) available to map, or ideally able to dump available mappable items? The logitech g13 daemon did this and was quite useful when mapping items.

Great work! Just picked this up today and was looking for this exact functionality :D


***edit** I was able to do this myself with evdev and a python script monitoring. However, once you map to a new key that is what evdev will report.

Incase it helps others.. Here is the list left to right of the Orbweaver. Can use these in a the translation list to set via python. 
       #top
'KEY_GRAVE':
'KEY_1':
'KEY_2':
'KEY_3':
'KEY_4':
#second
'KEY_TAB':
'KEY_Q':
'KEY_W':
'KEY_E':
'KEY_R':
#third
'KEY_CAPSLOCK':
'KEY_A':
'KEY_S':
'KEY_D':
'KEY_F':
#fourth
'KEY_LEFTSHIFT':
'KEY_Z':
'KEY_X':
'KEY_C':
'KEY_V':
#side
'KEY_LEFTALT':
'KEY_SPACE':
'KEY_UP':
'KEY_DOWN':
'KEY_LEFT':
'KEY_RIGHT':

**Comment by pmizio** [comment]
@jpyth and all future users 
You can detect keycodes using bellow command:
`xev | awk -F'[ )]+' '/^KeyPress/ { a[NR+2] } NR in a { printf "%-3s %s\n", $5-8, $8 }'`
Names aren't always right so you can check it in kernel header file with all keycodes by event number.
[https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h](https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h)
**Important:** Check on device with no keybindings set because when you checking binded device they send remapped keycodes not original.

**Comment by pmizio** [comment]
Hi guys how is progress on this I mean code review or another steps to progress on this pr?

**Comment by z3ntu** [comment]
Please rebase on master. I hope to review in the coming days.

**Comment by pmizio** [comment]
I have crazy month but branch is rebased now ;)

**Review by Akm0d — APPROVED** [review]

**Comment by shawnallen85** [comment]
Any update on getting this in? Thanks!!

**Comment by z3ntu** [comment]
Yes, sorry. It's another big PR and I'm waiting for #888 to be merged first before I start reviewing (and testing out the API etc) this PR.

From a quick overview I can say that I'm not really happy that the `'translations_get', 'translations_set'` methods are added per device and not for all devices - because most devices won't be touched again in the coming years so then either almost nobody will have this functionality or it's mass-added to all devices but then we can also make it a common attribute for all mice & keyboards.
Also in my opinion the daemon interface should have more type safety than a byte array - and dbus has a `DICT_ENTRY` type so it should be possible to map the byte array directly in the daemon (and then don't do much magic in the pylib).
And another thing, the docstrings for the new pylib methods are a bit non-descriptive, from looking at the docstring I have no idea what to expect from it (a dict but not at all what types are in there, what data is in these types, etc)

**Comment by chabad360** [comment]
A few questions and suggestions:

* Why not use the keymap system provided by [`keyboard.py`](https://github.com/openrazer/openrazer/blob/master/daemon/openrazer_daemon/keyboard.py) and [`key_event_management.py`](https://github.com/openrazer/openrazer/blob/master/daemon/openrazer_daemon/misc/key_event_management.py)? It seems that there would be some conflicts between keymap and this system (as @tildearrow noted).

* Would it be possible to have multiple keymaps along with different custom matrix effects for each one? Additionally, would the different keymaps be reflected by the profile LEDs?

* This doesn't appear to have any way to store the configuration (I suppose that would be added on top of #888)?

Thanks a lot for this desperately needed feature.

**Comment by z3ntu** [comment]
See #1124

**Comment by linuxisfreedom** [comment]
I've found a perfect solution to rebind all additional 12 keys of my Razer Naga Pro with the great tool "keyd".

**Pull request closed without merging**